src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRule.scala

```diff
@@ -106,10 +106,18 @@ object JoinIndexRule
    * @return true if supported. False if not.
    */
   private def isApplicable(l: LogicalPlan, r: LogicalPlan, condition: Expression): Boolean = {
+    // The given plan is eligible if it is supported and index has not been applied.
+    def isEligible(optRel: Option[FileBasedRelation]): Boolean = {
+      optRel.map(!RuleUtils.isIndexApplied(_)).getOrElse(false)
+    }
+
+    lazy val optLeftRel = RuleUtils.getRelation(spark, l)
+    lazy val optRightRel = RuleUtils.getRelation(spark, r)
+
     isJoinConditionSupported(condition) &&
-    RuleUtils.getLogicalRelation(l).isDefined && RuleUtils.getLogicalRelation(r).isDefined &&
-    isPlanLinear(l) && isPlanLinear(r) && !isPlanModified(l) && !isPlanModified(r) &&
-    ensureAttributeRequirements(l, r, condition)
+    isPlanLinear(l) && isPlanLinear(r) &&
```

Comment:
@apoorvedave1 we don't need `isPlanLinear` if `getRelation` implementation is as follows, right (collecting leaves and making sure it's only one)?:
```scala
  def getRelation(spark: SparkSession, plan: LogicalPlan): Option[FileBasedRelation] = {
    val provider = Hyperspace.getContext(spark).sourceProviderManager
    val leaves = plan.collectLeaves()
    if (leaves.size == 1 && provider.isSupportedRelation(leaves.head)) {
      Some(provider.getRelation(leaves.head))
    } else {
      None
    }
  }
```

Reply:
Great Thanks, i will update the join rule v2 to incorporate this as well